### PR TITLE
HDDS-9426. Calculate Exclusive size for deep cleaned snapshot's deleted directories.

### PR DESCRIPTION
**Pull request by aswinshakil**
## What changes were proposed in this pull request?

In #5339, We expanded the snapshot's `deletedDirTable` into `fileTable` to calculate the exclusive size for the files under the deleted directories. The problem with that approach is that if we alter the `fileTable` and `dirTable` of a snapshot, it would change the `snapDiff` result and yield undesirable results. For eg, Consider a deleted directory structure as below,
```
dir1 --- subdir1 --- file1 to file100
    | 
     ---- subdir2 --- file1 to file100

Instead of snapDiff saying DELETED dir1, It would say
DELETED subdir1
DELETED subdir2
DELETED subdir1/file1 to file100
DELETED subdir2/file1 to file100
```

In the new approach in this patch, We do an in-memory walk for each deleted directory and deep clean them, and also calculate the `exclusive size` for files when doing so. For each file, we do the same as #5301 

The reason we are doing a deep clean, There could be files under a directory trapped in between snapshots. Consider this case,
```
Create dir1
Create snapshot1
Create 1000 files under dir1 --> These files are trapped and not deleted unless the snapshot itself is deleted.
Delete dir1
Create snapshot2
```

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9426

## How was this patch tested?

Added Integration test.

### CODE REVIEW

**Review by hemantk-12 — COMMENTED** [review]
Thanks for the patch @aswinshakil

File: hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto [hunk 1]
```diff
@@ -1879,15 +1880,16 @@ message SnapshotMoveKeyInfos {
...
-  optional SnapshotProperty snapshotProperty = 1;
+  optional string snapshotKey = 1;
+  optional SnapshotSize snapshotSize = 2;
+  optional bool deepCleanedDeletedDir = 3;
+  optional bool deepCleanedDeletedKey = 4;
 }
 
-message SnapshotProperty {
-  optional string snapshotKey = 1;
+message SnapshotSize {
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/interface-client/src/main/proto/OmClientProtocol.proto:1892]: We don't need to, This is just for better encapsulation of sizes.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryService.java [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+import static org.apache.hadoop.ozone.OzoneConsts.OM_KEY_PREFIX;
+import static org.apache.hadoop.ozone.om.OmSnapshotManager.getSnapshotPrefix;
+import static org.apache.hadoop.ozone.om.helpers.SnapshotInfo.SnapshotStatus.SNAPSHOT_ACTIVE;
+import static org.apache.hadoop.ozone.om.request.file.OMFileRequest.getDirectoryInfo;
+
+/**
+ * Snapshot BG Service for deleted directory deep clean and exclusive size
+ * calculation for deleted directories.
+ */
+public class SnapshotDirectoryService extends AbstractKeyDeletingService {
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryService.java:75]: Makes sense to call `SnapshotDirectoryCleaningService`, I will update it

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryService.java [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+        SNAPSHOT_DIR_CORE_POOL_SIZE, serviceTimeout, ozoneManager, scmClient);
+    this.suspended = new AtomicBoolean(false);
+    this.exclusiveSizeMap = new HashMap<>();
+    this.exclusiveReplicatedSizeMap = new HashMap<>();
+  }
+
+  private boolean shouldRun() {
+    if (getOzoneManager() == null) {
+      // OzoneManager can be null for testing
+      return true;
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryService.java:99]: I should be based on suspended only when ozone manager is not null.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryService.java [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+    } catch (ServiceException e) {
+      LOG.error("Snapshot deep cleaning request failed. " +
+          "Will retry at next run.", e);
+    }
+  }
+
+  /**
+   * Stack node data for directory deep clean for snapshot.
+   */
+  @SuppressFBWarnings
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryService.java:499]: Yup removed it with changing the class to static

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryService.java [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+                OmKeyInfo>> deletedDirIterator = snapDeletedDirTable
+                .iterator(dbBucketKeyForDir)) {
+
+              while (deletedDirIterator.hasNext()) {
+                Table.KeyValue<String, OmKeyInfo> deletedDirInfo =
+                    deletedDirIterator.next();
+                String deletedDirKey = deletedDirInfo.getKey();
+
+                // Exit if it is out of the bucket scope.
+                if (!deletedDirKey.startsWith(dbBucketKeyForDir)) {
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryService.java:254]: Just added it for double-checking. I guess we can remove it with the prefix iterator.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java [hunk 6]
```diff
@@ -386,6 +386,18 @@ private OMConfigKeys() {
   public static final String OZONE_DIR_DELETING_SERVICE_INTERVAL_DEFAULT
       = "60s";
 
+  /**
+   * Configuration properties for Snapshot Directory Service.
+   */
+  public static final String OZONE_SNAPSHOT_DIRECTORY_SERVICE_INTERVAL =
+      "ozone.snapshot.directory.service.interval";
+  public static final String OZONE_SNAPSHOT_DIRECTORY_SERVICE_INTERVAL_DEFAULT
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java:394]: Yes, It runs every 24hrs. We can change the default values if we want.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+        throws IOException {
+    String seekFileInDB = metadataManager.getOzonePathKey(volumeId,
+        bucketInfo.getObjectID(),
+        parentInfo.getObjectID(), "");
+    List<BlockGroup> blocksForKeyDelete = new ArrayList<>();
+
+    Table<String, OmKeyInfo> fileTable = metadataManager.getFileTable();
+    try (TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+             iterator = fileTable.iterator()) {
+      iterator.seek(seekFileInDB);
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:402]: This implementation is from the `DirectoryDeletingService`.  Wouldn't `fileTable.iterator(seekFileInDB)` yield a prefix iterator? I think both should work in this case. Anyways I made the change here.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+              previousToPrevKeyTable);
+          seekDirInDB = currOmSnapshot.getMetadataManager()
+              .getOzonePathKey(volumeId, bucketInfo.getObjectID(),
+                  stackTop.getDirValue().getObjectID(), "");
+          directoryIterator.seek(seekDirInDB);
+        } else {
+          // When a leaf node is processed, we need come back in
+          // the call stack and process the next directories.
+          seekDirInDB = stackTop.getSubDirSeek();
+          directoryIterator.seek(seekDirInDB);
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:319]: Bloom filters may result in false positives, but in the end these results are merged. So it should yield correct results.

https://github.com/facebook/rocksdb/wiki/Prefix-Seek
```
Normally, when an iterator seek is executed, RocksDB needs to place the position at every sorted run (memtable, level-0 file, other levels, etc) to the seek position and merge these sorted runs. This can sometimes involve several I/O requests. It is also CPU heavy for decompression of several data blocks and other CPU overheads.
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+    try (
+        TableIterator<String, ? extends Table.KeyValue<String, OmDirectoryInfo>>
+            directoryIterator = snapDirTable.iterator(dbBucketKeyForDir)) {
+
+      while (!stackNodes.isEmpty()) {
+        StackNode stackTop = stackNodes.pop();
+        String seekDirInDB;
+        // First process all the files in the current directory
+        // and then do a DFS for directory.
+        if (StringUtils.isEmpty(stackTop.getSubDirSeek())) {
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:300]: The code above does level order traversal (BFS), the problem with the queue approach is that there will be more directories in a particular level than the entire depth of the directory tree. Consider a case there are 10K dirs in a particular level, keeping this in memory takes up significant JVM heap. But the depth of the directories is considerably small. So at max, we keep the depth of the tree in the stack. 

Doing a BFS approach will indeed make this whole traversal very simple. The only problem is the in-memory queue we maintain can become huge.

[the remark this answers is not in the document]
Comment (hemantk-12) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:298]: I see your point.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (hemantk-12) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:317]: Seek with prefix caused HDDS-7871.

**Review by hemantk-12 — COMMENTED** [review]
Overall changes look good to me. Left some minor and cosmetic comments.

We can merge this once you resolve them.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 10]
```diff
@@ -311,10 +311,12 @@ private void iterateDirectoryTree(
                   stackTop.getDirValue().getObjectID(), "");
           directoryIterator.seek(seekDirInDB);
         } else {
-          // When a leaf node is processed, we need come back in
+          // When a leaf node is processed, we need to come back in
           // the call stack and process the next directories.
           seekDirInDB = stackTop.getSubDirSeek();
           directoryIterator.seek(seekDirInDB);
+          // We need skip to the next sub-directory because we already
```
Comment (hemantk-12) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:318]: nit: either `We need to skip ...` or `We skip ...`.

**Review by hemantk-12 — COMMENTED** [review]
Thanks for updating it @aswinshakil 

LGTM except the seek and next as in comment: https://github.com/apache/ozone/pull/5579/files#r1413457354

**Review by hemantk-12 — APPROVED** [review]
LGTM.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java [hunk 11]
```diff
@@ -469,6 +469,98 @@ public long optimizeDirDeletesAndSubmitRequest(long remainNum,
...
+      OmKeyInfo keyInfoPrevSnapshot = getPreviousSnapshotKeyName(
+          keyInfo, bucketInfo, volumeId,
+          snapRenamedTable, previousKeyTable);
+      OmKeyInfo keyInfoPrevToPrevSnapshot = getPreviousSnapshotKeyName(
+          keyInfoPrevSnapshot, bucketInfo, volumeId,
+          prevRenamedTable, previousToPrevKeyTable);
+      // If the previous to previous snapshot doesn't
+      // have the key, then it is exclusive size for the
+      // previous snapshot.
+      if (keyInfoPrevToPrevSnapshot == null) {
```
Comment (swamirishi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java:522]: We are supposed to look for the version diff, in case of object versioning. Exclusive size would be the size of the versions not found in the previous to previous and found in the previous snapshot.

**Review by swamirishi — CHANGES_REQUESTED** [review]
@aswinshakil Thanks for the patch. I have a few major review comments after going through the PR so far. I haven't gone through the entire code and logic yet. These are some of the review comments you can address in the meanwhile I review the rest of the PR.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+    stackNodes.add(topLevelDir);
+
+    try (
+        TableIterator<String, ? extends Table.KeyValue<String, OmDirectoryInfo>>
+            directoryIterator = snapDirTable.iterator(dbBucketKeyForDir)) {
+
+      while (!stackNodes.isEmpty()) {
+        StackNode stackTop = stackNodes.pop();
+        String seekDirInDB;
+        // First process all the files in the current directory
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:296]: I was suggesting to go with DFS previously. We had a discussion where you suggested DFS instead of BFS because of in-memory footprint of the background process.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+              currSnapInfo, snapChainManager, omSnapshotManager);
+          SnapshotInfo previousToPrevSnapshot = null;
+
+          Table<String, OmKeyInfo> previousKeyTable = null;
+          Table<String, String> prevRenamedTable = null;
+          ReferenceCounted<IOmMetadataReader, SnapshotCache>
+              rcPrevOmSnapshot = null;
+
+          if (previousSnapshot != null) {
+            rcPrevOmSnapshot = omSnapshotManager.checkForSnapshot(
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:186]: Nice catch. Added it. Thanks!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+          // We need skip to the next sub-directory because we already
+          // processed the current sub-directory in the previous run.
+          if (directoryIterator.hasNext()) {
+            directoryIterator.next();
+          } else {
+            continue;
+          }
+        }
+
+        if (directoryIterator.hasNext()) {
```
[the remark this answers is not in the document]
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:327]: We need the current directory reference to add the next sub-dir to the stack. So it is simpler to do it in the same iteration than the next iteration.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java (hunk 11 above)
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java:522]: Updated it.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:303]: yeah agreed in memory footprint will be higher for bfs in case of high fan out.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+    topLevelDir.setDirKey(dirPathDbKey);
+    topLevelDir.setDirValue(omDeletedDirectoryInfo);
+    stackNodes.add(topLevelDir);
+
+    try (
+        TableIterator<String, ? extends Table.KeyValue<String, OmDirectoryInfo>>
+            directoryIterator = snapDirTable.iterator(dbBucketKeyForDir)) {
+
+      while (!stackNodes.isEmpty()) {
+        StackNode stackTop = stackNodes.pop();
```
[the remark this answers is not in the document]
Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:301]: Why do we have to pop and push. We can do peek and pop from the stack only if required.

[the remark this answers is not in the document]
Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:301]:
```suggestion
        if (StringUtils.isEmpty(stackTop.getSubDirSeek())) {
          processFilesUnderDir(previousSnapshot,
              previousToPrevSnapshot,
              volumeId,
              bucketInfo,
              stackTop.getDirValue(),
              currOmSnapshot.getMetadataManager(),
              snapRenamedTable,
              previousKeyTable,
              prevRenamedTable,
              previousToPrevKeyTable);
          seekDirInDB = currOmSnapshot.getMetadataManager()
              .getOzonePathKey(volumeId, bucketInfo.getObjectID(),
                  stackTop.getDirValue().getObjectID(), "");
          stackTop.setSubDirSeek(seekDirInDB);
        } else {
          directoryIterator.seek(stackTop.getSubDirSeek() + "\0");
          if (directoryIterator.hasNext()) {
            Table.KeyValue<String, OmDirectoryInfo> deletedSubDirInfo =
                directoryIterator.next();
            String deletedSubDirKey = deletedSubDirInfo.getKey();

            String prefixCheck = currOmSnapshot.getMetadataManager()
                .getOzoneDeletePathDirKey(stackTop.getSubDirSeek());
            // Exit if it is out of the sub dir prefix scope.
            if (!deletedSubDirKey.startsWith(prefixCheck)) {
              // Add exit condition.
              stackNodes.pop();
            } else {
              stackTop.setSubDirSeek(deletedSubDirKey);
              StackNode nextSubDir = new StackNode();
              nextSubDir.setDirKey(deletedSubDirInfo.getKey());
              nextSubDir.setDirValue(deletedSubDirInfo.getValue());
              stackNodes.push(nextSubDir);  
            }
          } else {
            stackNodes.pop();
          }
        }
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 16]
```diff
@@ -0,0 +1,530 @@
...
+    topLevelDir.setDirKey(dirPathDbKey);
+    topLevelDir.setDirValue(omDeletedDirectoryInfo);
+    stackNodes.push(topLevelDir);
+
+    try (
+        TableIterator<String, ? extends Table.KeyValue<String, OmDirectoryInfo>>
+            directoryIterator = snapDirTable.iterator(dbBucketKeyForDir)) {
+
+      while (!stackNodes.isEmpty()) {
+        StackNode stackTop = stackNodes.pop();
```
Comment (swamirishi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:301]:
```suggestion
        StackNode stackTop = stackNodes.peek();
        String seekDirInDB;
        // First process all the files in the current directory
        // and then do a DFS for directory.
        if (StringUtils.isEmpty(stackTop.getSubDirSeek())) {
          processFilesUnderDir(previousSnapshot,
              previousToPrevSnapshot,
              volumeId,
              bucketInfo,
              stackTop.getDirValue(),
              currOmSnapshot.getMetadataManager(),
              snapRenamedTable,
              previousKeyTable,
              prevRenamedTable,
              previousToPrevKeyTable);
          seekDirInDB = currOmSnapshot.getMetadataManager()
              .getOzonePathKey(volumeId, bucketInfo.getObjectID(),
                  stackTop.getDirValue().getObjectID(), "");
          stackTop.setSubDirSeek(seekDirInDB);
        } else {
          directoryIterator.seek(stackTop.getSubDirSeek() + "\0");
          if (directoryIterator.hasNext()) {
            Table.KeyValue<String, OmDirectoryInfo> deletedSubDirInfo =
                directoryIterator.next();
            String deletedSubDirKey = deletedSubDirInfo.getKey();

            String prefixCheck = currOmSnapshot.getMetadataManager()
                .getOzoneDeletePathDirKey(stackTop.getSubDirSeek());
            // Exit if it is out of the sub dir prefix scope.
            if (!deletedSubDirKey.startsWith(prefixCheck)) {
              // Add exit condition.
              stackNodes.pop();
            } else {
              stackTop.setSubDirSeek(deletedSubDirKey);
              StackNode nextSubDir = new StackNode();
              nextSubDir.setDirKey(deletedSubDirInfo.getKey());
              nextSubDir.setDirValue(deletedSubDirInfo.getValue());
              stackNodes.push(nextSubDir);
            }
          } else {
            stackNodes.pop();
          }
        }
```

Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:301]: Does this implementation look simpler? I have added "\0" at the end of subDir seek, so it will be the string just greater than the seeked string and next string will be iterated. I also changed the implementation to set the dirSeek after coming across the node the first time i.e. cleaning up all files in dir. Next iteration will start iterating through the sub dirs & add sub dir to stack one by one. One iteration at ta time.

**Review by swamirishi — COMMENTED** [review]
Added some suggestion. You may take it if it is helpful

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java (hunk 16 above)
Comment (aswinshakil) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:301]: This looks great. Thanks for the detailed suggestion.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java (hunk 11 above)
Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java:522]: where is the version size being checked?

Comment (swamirishi) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/AbstractKeyDeletingService.java:522]: One version of the object would correspond to only one snapshot we need to check which version this is. It could be the case the key was modified and we are deleting previous versions or we are keeping all object versions.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java [hunk 17]
```diff
@@ -0,0 +1,515 @@
...
+              previousKeyTable,
+              prevRenamedTable,
+              previousToPrevKeyTable);
+          // Format : /volId/bucketId/parentId/
+          String seekDirInDB = currOmSnapshot.getMetadataManager()
+              .getOzonePathKey(volumeId, bucketInfo.getObjectID(),
+                  stackTop.getDirValue().getObjectID(), "");
+          stackTop.setSubDirSeek(seekDirInDB);
+        } else {
+          // Adding \0 to seek the next greater element.
```
Comment (swamirishi) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:320]: @hemantk-12 Can you confirm if this is a correct approach. @aswinshakil it would be great to add this edge case in the unit test if it is not there.

Comment (hemantk-12) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDirectoryCleaningService.java:320]: It looks good to me. As you said, we have to test it thoroughly.

**Review by swamirishi — APPROVED** [review]
@aswinshakil thanks for the patch. I have raised a follow up jira for fixing the exclusive size calculation for buckets with object versioning enabled. https://issues.apache.org/jira/browse/HDDS-10051

**Comment by aswinshakil** [comment]
Thank you @hemantk-12  and @swamirishi  for the reviews!
